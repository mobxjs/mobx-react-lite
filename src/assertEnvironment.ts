import { spy } from "mobx"
import { useState } from "react"

if (!useState) {
    throw new Error("mobx-react-lite requires React with Hooks support")
}
if (!spy) {
    throw new Error("mobx-react-lite requires mobx at least version 4 to be available")
}
