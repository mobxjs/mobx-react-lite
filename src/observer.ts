import { forwardRef, memo } from "react"
import { isUsingStaticRendering } from "./staticRendering"
import { useObserver } from "./useObserver"
import hoistStatics from "hoist-non-react-statics"

export interface IObserverOptions {
    readonly forwardRef?: boolean
}

export function observer<P extends object, TRef = {}>(
    baseComponent: React.RefForwardingComponent<TRef, P>,
    options: IObserverOptions & { forwardRef: true }
): React.MemoExoticComponent<
    React.ForwardRefExoticComponent<React.PropsWithoutRef<P> & React.RefAttributes<TRef>>
>
export function observer<P extends object>(
    baseComponent: React.FunctionComponent<P>,
    options?: IObserverOptions
): React.NamedExoticComponent<P>
// n.b. base case is not used for actual typings or exported in the typing files
export function observer<P extends object, TRef = {}>(
    baseComponent: React.RefForwardingComponent<TRef, P>,
    options?: IObserverOptions
) {
    // The working of observer is explaind step by step in this talk: https://www.youtube.com/watch?v=cPF4iBedoF0&feature=youtu.be&t=1307
    if (isUsingStaticRendering()) {
        return baseComponent
    }

    const realOptions = {
        forwardRef: false,
        ...options
    }

    const baseComponentName = baseComponent.displayName || baseComponent.name

    const wrappedComponent = (props: P, ref: React.Ref<TRef>) => {
        return useObserver(() => baseComponent(props, ref), baseComponentName)
    }

    // memo; we are not intested in deep updates
    // in props; we assume that if deep objects are changed,
    // this is in observables, which would have been tracked anyway
    let memoComponent
    if (realOptions.forwardRef) {
        // we have to use forwardRef here because:
        // 1. it cannot go before memo, only after it
        // 2. forwardRef converts the function into an actual component, so we can't let the baseComponent do it
        //    since it wouldn't be a callable function anymore
        memoComponent = memo(forwardRef(wrappedComponent))
    } else {
        memoComponent = memo(wrappedComponent)
    }

    copyStaticProperties(baseComponent as any, memoComponent as any)
    memoComponent.displayName = baseComponentName

    return memoComponent
}

function copyStaticProperties(base: React.FunctionComponent, target: React.FunctionComponent) {
    hoistStatics(target, base) // See #32
    // propTypes and defaultProps need to be hoisted as well since we wrap and invoke the component as function, not as type
    if (base.propTypes) {
        target.propTypes = base.propTypes
    }
    if (base.defaultProps) {
        target.defaultProps = base.defaultProps
    }
}
