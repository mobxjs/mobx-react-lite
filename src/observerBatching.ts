import { configure } from "mobx"
import { getGlobal, getSymbol } from "./utils"

const observerBatchingConfiguredSymbol = getSymbol("observerBatching")

export function defaultNoopBatch(callback: () => void) {
    callback()
}

export function observerBatching(reactionScheduler: any) {
    if (!reactionScheduler) {
        reactionScheduler = defaultNoopBatch
        if ("production" !== process.env.NODE_ENV) {
            console.warn(
                "[MobX] Failed to get unstable_batched updates from react-dom / react-native"
            )
        }
    }
    configure({ reactionScheduler })
    getGlobal()[observerBatchingConfiguredSymbol] = true
}

export const isObserverBatched = () => !!getGlobal()[observerBatchingConfiguredSymbol]
