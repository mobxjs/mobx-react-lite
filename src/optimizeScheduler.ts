import { configure } from "mobx"

interface TBatchedUpdated {
    batchedUpdates<A, B>(callback: (a: A, b: B) => any, a: A, b: B): void
    batchedUpdates<A>(callback: (a: A) => any, a: A): void
    batchedUpdates(callback: () => any): void
}

export const optimizeScheduler = (reactionScheduler: TBatchedUpdated) => {
    if (typeof reactionScheduler === "function") {
        configure({ reactionScheduler })
    }
}

export const deoptimizeScheduler = () => {
    configure({ reactionScheduler: undefined })
}
