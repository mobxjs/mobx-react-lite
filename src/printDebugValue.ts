import { getDependencyTree, Reaction } from "mobx"

export function printDebugValue(v: React.MutableRefObject<Reaction | null>) {
    if (!v.current) {
        return "<unknown>"
    }
    return getDependencyTree(v.current)
}
