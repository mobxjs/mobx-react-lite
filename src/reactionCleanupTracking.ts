import { Reaction } from "mobx"

export interface IReactionTracking {
    /** The Reaction created during first render, which may be leaked */
    reaction: Reaction
    /**
     * The time (in ticks) at which point we should dispose of the reaction
     * if this component hasn't yet been fully mounted.
     */
    cleanAt: number

    /**
     * Whether the component has yet completed mounting (for us, whether
     * its useEffect has run)
     */
    mounted?: boolean

    /**
     * Whether the observables that the component is tracking changed between
     * the first render and the first useEffect.
     */
    changedBeforeMount?: boolean
}

export function createTrackingData(reaction: Reaction) {
    const trackingData: IReactionTracking = {
        cleanAt: Date.now() + CLEANUP_LEAKED_REACTIONS_AFTER_MILLIS,
        reaction
    }
    return trackingData
}

/**
 * The minimum time before we'll clean up a Reaction created in a render
 * for a component that hasn't managed to run its effects. This needs to
 * be big enough to ensure that a component won't turn up and have its
 * effects run without being re-rendered.
 */
export const CLEANUP_LEAKED_REACTIONS_AFTER_MILLIS = 10_000

/**
 * The frequency with which we'll check for leaked reactions.
 */
export const CLEANUP_TIMER_LOOP_MILLIS = 10_000

/**
 * Reactions created by components that have yet to be fully mounted.
 */
const uncommittedReactionRefs: Set<React.MutableRefObject<IReactionTracking | null>> = new Set()

/**
 * Latest 'uncommitted reactions' cleanup timer handle.
 */
let reactionCleanupHandle: ReturnType<typeof setTimeout> | undefined

function ensureCleanupTimerRunning() {
    if (reactionCleanupHandle === undefined) {
        reactionCleanupHandle = setTimeout(cleanUncommittedReactions, CLEANUP_TIMER_LOOP_MILLIS)
    }
}

export function scheduleCleanupOfReactionIfLeaked(
    ref: React.MutableRefObject<IReactionTracking | null>
) {
    uncommittedReactionRefs.add(ref)

    ensureCleanupTimerRunning()
}

export function recordReactionAsCommitted(
    reactionRef: React.MutableRefObject<IReactionTracking | null>
) {
    uncommittedReactionRefs.delete(reactionRef)
}

/**
 * Run by the cleanup timer to dispose any outstanding reactions
 */
function cleanUncommittedReactions() {
    reactionCleanupHandle = undefined

    // Loop through all the candidate leaked reactions; those older
    // than CLEANUP_LEAKED_REACTIONS_AFTER_MILLIS get tidied.

    const now = Date.now()
    uncommittedReactionRefs.forEach(ref => {
        const tracking = ref.current
        if (tracking) {
            if (now >= tracking.cleanAt) {
                // It's time to tidy up this leaked reaction.
                tracking.reaction.dispose()
                ref.current = null
                uncommittedReactionRefs.delete(ref)
            }
        }
    })

    if (uncommittedReactionRefs.size > 0) {
        // We've just finished a round of cleanups but there are still
        // some leak candidates outstanding.
        ensureCleanupTimerRunning()
    }
}

/* istanbul ignore next */
/**
 * Only to be used by test functions; do not export outside of mobx-react-lite
 */
export function forceCleanupTimerToRunNowForTests() {
    // This allows us to control the execution of the cleanup timer
    // to force it to run at awkward times in unit tests.
    if (reactionCleanupHandle) {
        clearTimeout(reactionCleanupHandle)
        cleanUncommittedReactions()
    }
}

/* istanbul ignore next */
export function resetCleanupScheduleForTests() {
    if (reactionCleanupHandle) {
        clearTimeout(reactionCleanupHandle)
        reactionCleanupHandle = undefined
    }
    uncommittedReactionRefs.clear()
}
