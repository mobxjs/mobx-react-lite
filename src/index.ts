import "./assertEnvironment"

export { useObservable } from "./useObservable"
export { useComputed } from "./useComputed"
export { useDisposable } from "./useDisposable"
export { isUsingStaticRendering, useStaticRendering } from "./staticRendering"
export { observer, IObserverOptions } from "./observer"
export { useObserver, ForceUpdateHook, IUseObserverOptions } from "./useObserver"
export { Observer } from "./ObserverComponent"
export { useForceUpdate } from "./utils"
