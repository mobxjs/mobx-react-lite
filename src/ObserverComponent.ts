import { ReactElement } from "react"

import { observer } from "./observer"

interface IObserverProps {
    children?(): ReactElement<any>
    render?(): ReactElement<any>
}

function ObserverComponent({ children, render }: IObserverProps) {
    const component = children || render
    if (typeof component === "undefined") {
        return null
    }
    return component()
}
ObserverComponent.propTypes = {
    children: ObserverPropsCheck,
    render: ObserverPropsCheck
}
ObserverComponent.displayName = "Observer"

export const Observer = observer(ObserverComponent)

function ObserverPropsCheck(
    props: { [k: string]: any },
    key: string,
    componentName: string,
    location: any,
    propFullName: string
) {
    const extraKey = key === "children" ? "render" : "children"
    const hasProp = typeof props[key] === "function"
    const hasExtraProp = typeof props[extraKey] === "function"
    if (hasProp && hasExtraProp) {
        return new Error(
            "MobX Observer: Do not use children and render in the same time in`" + componentName
        )
    }

    if (hasProp || hasExtraProp) {
        return null
    }
    return new Error(
        "Invalid prop `" +
            propFullName +
            "` of type `" +
            typeof props[key] +
            "` supplied to" +
            " `" +
            componentName +
            "`, expected `function`."
    )
}
