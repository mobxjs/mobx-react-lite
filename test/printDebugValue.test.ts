import { $mobx, autorun, observable } from "mobx"
import { printDebugValue } from "../src/printDebugValue"

test("printDebugValue", () => {
    const money = observable({
        euro: 10,
        get pound() {
            return this.euro / 1.15
        }
    })

    const disposer = autorun(() => {
        const { euro, pound } = money
        if (euro === pound) {
            // tslint:disable-next-line: no-console
            console.log("Weird..")
        }
    })

    const value = (disposer as any)[$mobx]

    expect(printDebugValue(value)).toMatchSnapshot()

    disposer()

    expect(printDebugValue(value)).toMatchSnapshot()
})
