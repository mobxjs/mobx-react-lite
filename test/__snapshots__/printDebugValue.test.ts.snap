// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printDebugValue 1`] = `
Object {
  "dependencies": Array [
    Object {
      "name": "ObservableObject@1.euro",
    },
    Object {
      "name": "ObservableObject@1.pound?",
    },
    Object {
      "dependencies": Array [
        Object {
          "name": "ObservableObject@1.euro",
        },
      ],
      "name": "ObservableObject@1.pound",
    },
  ],
  "name": "Autorun@3",
}
`;

exports[`printDebugValue 2`] = `
Object {
  "name": "Autorun@3",
}
`;
